crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+                    // will yield `None` and we can terminate the event loop
+                    let message = if let Some(message) = maybe_message {
+                        message
+                    } else {
+                        break;
+                    };
+
+                    self.handle_message(message);
+                }
+                peer_event = peer_events.recv() => {
+                    self.handle_peer_event(peer_event);
+                },
+                Some(task_result) = self.tasks.join_next() => {
+                    task_result.unwrap();
+                },
+                Some(sync_result) = self.sync_tasks.join_next() => {
+                    let maybe_checkpoint = sync_result.unwrap();
+                    // On a successful sync, gossip out the checkpoint we synced to
+                    if let Some(checkpoint) = maybe_checkpoint {
+                        self.gossip_checkpoint(checkpoint);
+                    }
+                }
+            }
+
+            self.maybe_start_sync_job();
+        }
+
+        info!("State-Synchronizer ended");
+    }
+
+    fn handle_message(&mut self, message: StateSyncMessage) {
+        match message {
+            StateSyncMessage::StartSyncJob => self.maybe_start_sync_job(),
+            StateSyncMessage::VerifiedCheckpoint(checkpoint) => self.handle_checkpoint(checkpoint),
+        }
+    }
+
+    // Handle a checkpoint that we recieved from consensus
+    fn handle_checkpoint(&mut self, checkpoint: Box<VerifiedCheckpoint>) {
+        let (next_sequence_number, previous_digest) = {
+            let latest_checkpoint = self.store.get_latest_checkpoint();
+
+            // If this is an older checkpoint, that's weird but ok, just ignore it
+            if latest_checkpoint.as_ref().map(|x| x.sequence_number())
+                >= Some(checkpoint.sequence_number())
+            {
+                return;
+            }
+
+            let next_sequence_number = latest_checkpoint
+                .as_ref()
+                .map(|x| x.sequence_number().saturating_add(1))
+                .unwrap_or(0);
+            let previous_digest = latest_checkpoint.map(|x| x.digest());
+            (next_sequence_number, previous_digest)
+        };
+
+        // If this is exactly the next checkpoint then lets insert it and gossip it out
+        if checkpoint.sequence_number() == next_sequence_number
+            && checkpoint.previous_digest() == previous_digest
+        {
+            let checkpoint = *checkpoint;
+            self.store.insert_checkpoint(checkpoint.clone());
+
+            // We don't care if no one is listening as this is a broadcast channel
+            let _ = self.checkpoint_event_sender.send(checkpoint.clone());
+
+            self.gossip_checkpoint(checkpoint);
+        } else {
+            // Otherwise shove it in the quarentined checkpoints and we can try to sync the missing
```

Comment:
can you say "unprocessed" here instead of quarantined to match the terminology used elsewhere in the code?